compiler/src/main/scala/org/scalajs/core/compiler/JSEncoding.scala

```diff
@@ -263,4 +265,19 @@ trait JSEncoding extends SubComponent { self: GenJSCode =>
     if (js.isKeyword(name) || name(0).isDigit || name(0) == '$')
       "$" + name
     else name
+
+
+  private def outerFieldSymSyffix(sym: Symbol): String = {
+    def allOutersInSym(acc: Set[Symbol], sym: Symbol): Set[Symbol] = {
+      val outersInSym = sym.info.members.filter(_.isOuterField).toSet
+      sym.parentSymbols.foldLeft(acc ++ outersInSym)(allOutersInSym)
+    }
+
+    val outers = allOutersInSym(Set.empty, sym.owner).toArray.sortBy(_.toString)
+    val index = outers.indexOf(sym)
+    assert(index >= 0)
+
+    if (index == 0) "f"
+    else "f$" + index
```

Comment:
In this new version, as long as there are no clashes the name will still have the `"f"` suffix (i.e. it will be binary compatible). In the other case the code did not compile in the examples that I tried and therefore binary compatibility should not be an issue. 
@sjrd @gzm0 is there a convoluted way to have two `$outer` fields that have the same type but do not refer to the same outer object?